src/services/download/sentry.rs

```diff
@@ -10,91 +10,86 @@ use std::sync::Arc;
 use url::Url;
 
 use super::{DownloadError, DownloadErrorKind, DownloadStatus, USER_AGENT};
 use crate::sources::{FileType, SentryFileId, SentrySourceConfig, SourceFileId};
 use crate::types::ObjectId;
+use crate::utils::futures::{delay, TimedTaskExt};
 
 lazy_static::lazy_static! {
     static ref CLIENT_CONNECTOR: Addr<ClientConnector> = ClientConnector::default().start();
     static ref SENTRY_SEARCH_RESULTS: Mutex<lru::LruCache<SearchQuery, (Instant, Vec<SearchResult>)>> =
         Mutex::new(lru::LruCache::new(100_000));
 }
 
-pub fn download_source(
+async fn start_request(
+    source: &SentrySourceConfig,
+    url: &Url,
+) -> Result<ClientResponse, SendRequestError> {
+    // The timeout is for the entire HTTP download *including* the response stream itself,
+    // in contrast to what the Actix-Web docs say. We have tested this manually.
+    //
+    // The intent is to disable the timeout entirely, but there is no API for that.
+    client::get(url)
+        .with_connector((*CLIENT_CONNECTOR).clone())
+        .header("User-Agent", USER_AGENT)
+        .header("Authorization", format!("Bearer {}", &source.token))
+        .timeout(Duration::from_secs(9999))
+        .finish()
+        .unwrap()
+        .send()
+        .compat()
+        .await
+}
+
+pub async fn download_source(
     source: Arc<SentrySourceConfig>,
     file_id: SentryFileId,
     destination: PathBuf,
-) -> Box<dyn Future<Item = DownloadStatus, Error = DownloadError>> {
+) -> Result<DownloadStatus, DownloadError> {
     let download_url = source.download_url(&file_id);
-
     log::debug!("Fetching debug file from {}", download_url);
-    let token = &source.token;
-    let response = clone!(token, download_url, || {
-        client::get(&download_url)
-            .with_connector((*CLIENT_CONNECTOR).clone())
-            .header("User-Agent", USER_AGENT)
-            .header("Authorization", format!("Bearer {}", token))
-            // This timeout is for the entire HTTP download *including* the response stream
-            // itself, in contrast to what the Actix-Web docs say. We have tested this
-            // manually.
-            //
-            // The intent is to disable the timeout entirely, but there is no API for that.
-            .timeout(Duration::from_secs(9999))
-            .finish()
-            .unwrap()
-            .send()
-    });
-
-    let response = Retry::spawn(
-        ExponentialBackoff::from_millis(10).map(jitter).take(3),
-        response,
-    );
-
-    let response = response.map_err(|e| match e {
-        tokio_retry::Error::OperationError(e) => e,
-        e => panic!("{}", e),
-    });
+    let mut backoff = ExponentialBackoff::from_millis(10).map(jitter).take(3);
+    let response = loop {
+        let result = start_request(&source, &download_url).await;
+        match backoff.next() {
+            Some(duration) if result.is_err() => delay(duration).await,
+            _ => break result,
+        }
+    };
```

Comment:
As often as we have this now, how about we create a `retry` utility for the downloaders? Doesn't have to be very configurable, I'm thinking:

```rust
async fn retry<T, E, F, C>(f: C) -> Result<T, E>
where
  C: FnMut() -> F,
  F: Future<Item = Result<T, E>>;
```

I believe that it should work just fine with `FnMut`, but not 100% sure.